docs: 16193 Lifecycle management for the State (#16592)
hashgraph/hedera-services: platform-sdk/docs/proposals/state-lifecycle/proposal.md

These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should offer all necessary mechanisms for the platform to interact with the state and its lifecycle.

Therefore, `swirlds-state-api` needs a `StateLifecycleManager` interface with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`  that will have the following responsibilities:
- Create/load an initial state as a starting point. If no snapshot is available, it should create a new state. 
- Managing references to the latest immutable state and the latest mutable state.
- Restrict mutability to a single state. Specifically, only one state object should be mutable at a time. To achieve this, we need to use a combination of `StateLifecycleManager.copy` method and `NonCopyableState` class.
- Load snapshots from the disk. A state loaded from a snapshot must be immutable.

The interface will look something like this:

```java
public interface StateLifecycleManager {
  /**
   * Get the latest immutable state.
   */
  State getLatestImmutableState();

  /**
   * Get the mutable state.
   */

>   State getMutableState();

Reviewer: A question on this came up in Block Node design discussion.
If I have a `mutable` state, make a large set of changes, then decide to _discard_ those changes, can I discard the `mutable` state, and get a _new_ `mutable` state from the `last immutable` state?

Functionally (using this API), can I efficiently revert back to the last immutable state and then start making changes again?

The use case here is if I make a bunch of state changes in Block Node, but then something causes the block stream to stop and have to retry the current block (network interruption, failed block proof, etc...), I need a way go back to the starting state and begin applying the same block again.

Author: >If I have a mutable state, make a large set of changes, then decide to discard those changes, can I discard the mutable state, and get a new mutable state from the last immutable state?

If we use the pattern established for the `MerkleNode` instances and their `copy` methods, we can't. This is one-way street. That is, once you create a copy of a mutable instance it becomes immutable, and only the copy is mutable. The framework doesn't allow to create a copy of an immutable instance
`com.swirlds.state.merkle.MerkleStateRoot#copy`:

```
   public T copy() {
        throwIfImmutable();
        throwIfDestroyed();
        setImmutable(true);
        return copyingConstructor();
    }
```

>Functionally (using this API), can I efficiently revert back to the last immutable state and then start making changes again?

I think we can support this use case but it require a tiny bit more effort as it deviates from the normal usage of a copy method.

Reviewer: We'll need that revert-to-last-immutable capability in Block Node to handle when a block fails validation or other interruptions, so we would very much appreciate if that could be included in the design for this API.

Author: Done

Reviewer: Thank you.

# Lifecycle management for the State

## Summary

This proposal describes a possible implementation of state lifecycle management that could work equally well for both the Consensus Node and the Block Node.

|      Metadata      |                 Entities                 |
|--------------------|------------------------------------------|
| Designers          | [@imalygin](https://github.com/imalygin) |
| Functional Impacts | Consensus Node, Block Node               |

## Purpose and Context

There is an initiative to create self-contained State modules (`swirlds-state-api` and `swirlds-state-impl`) that would have minimal dependencies, and specifically, they should not depend on `swirlds-platform-core`. Part of this initiative is the implementation of a state lifecycle management mechanism that could be used by both the Consensus Node and the Block Node.

### Requirements

- Add an interface representing state lifecycle management and its implementation.
- Update the current State API to be coherent with this State Lifecycle API.
- Implement this set of APIs and migrate the existing code to use it.

> ## Design & Architecture

Reviewer 2: I find myself wanting some summary of what the new functionality is at a level between the `Purpose and Context` and the `Design & Architecture` levels. Something that describes what this new feature will do before getting into the details like methods. What is meant by "lifecycle management" now that the client must maintain references?

Reviewer 2: Somewhere, I'm not sure where but maybe here, it would be good to describe how the states flow through the CN along its lifecycle at a high level and how things will change with this new State API. For example, the fact that the CN will invoke the State Module to load a state from disk and to make fast copies when a round is done being handled, but still need to hold onto state objects with reservations to collect signatures. Then the equivalent for BN.

Author: I updated an existing diagram mapping the State method names to the part of the lifecycle. I also added a diagram that depicts how the final result is going to look like once we add `StateLifecycleManager`

### Reservation count mechanism

> This mechanism is already implemented as a part of the Merkle tree implementation. However, it's importnant to mention it here

Reviewer 2: ```suggestion
This mechanism is already implemented as a part of the Merkle tree implementation. However, it's important to mention it here
```

Author: I didn't just fixed the typo, I reworded the whole sentence.

to provide important context for understanding how exactly the state lifecycle management will work.

Every Merkle tree node implements the following interface:

> ```java

Reviewer 2: I think putting this code block here is not necessary, and is in fact distracting. It's large and takes up a lot of space, and breaks up my mental flow. I think a few sentences describing the basics of the reservations would be much better.

Author: Fair enough. I'll remove the code and replace it with the summary of the idea.

Author: Done.

     * if this object only has an implicit reference (i.e. {@link #getReservationCount()} returns
     * {@link #IMPLICIT_REFERENCE_COUNT}). An object has an implicit reference immediately after it is constructed
     * but before {@link #reserve()} has been called the first time. If called with an implicit reference, this
     * object will be destroyed.
     * </p>
     *
     * @return true if this call to release() caused the object to become destroyed
     * @throws com.swirlds.common.exceptions.ReferenceCountException
     * 		this object has already been fully released and destroyed
     */
    @Override
    boolean release();

    /**
     * Get the total number of times {@link #reserve()} has been called minus the number of times {@link #release()}
     * has been called. Will return {@link #IMPLICIT_REFERENCE_COUNT} if {@link #reserve()} has never been called,
     * or {@link #DESTROYED_REFERENCE_COUNT} if this object has been fully released.
     *
     * @return
     */
    int getReservationCount();
}

```

We're going to rely on this mechanism to manage the number of states that are kept in memory. 

> **Important**: It's up to the client code to make sure that the state is released once it's no longer needed.

Reviewer 2: Will the State Module ensure that a state is at least hashed before it is garbage collected even if it's reservations are at 0? Or is it entirely on the client to maintain reservations in all cases?

Author: I see two possible approaches to that: 
- It may be entirely up to client to maintain reservations and make sure that they are acquired and released properly. In this case, `StateLifecycleManager` will not track if the state is hashed or not. 
- Reservation maintenance may be responsibility of `StateLifecycleManager`, and direct reservation of `State` instance should not be allowed. In this case, `StateLifecycleManager` has more control over `State` and can make sure that it's hashed before it's released but in this case we would have to maintain a list of `State` instances in memory. 

I'm leaning towards option 1.

Once the reservation count reaches 0, the state will be eligible for garbage collection.

### Java classes

- `com.swirlds.state.State` has the following lifecycle-related methods:
  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
  - `void computeHash()`: computes the hash of the immutable state.
  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.

These methods correspond to the state lifecycle as follows:

> [![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)

Reviewer 2: This diagram and the classes now seem so to be a bit contradictory. Is describes a `State` class that has methods like `copy()` and `computeHash()`, yet these operations are described/depicted as part of the lifecycle of the state. Then there is the `StateLifecycleManager` that has `copy()`. In all cases, `copy()` can only be called on the mutable state. So we need to be very clear about which operations are part of the lifecycle and put those in the lifecycle manager and not duplicate the copy method.

Reviewer 2: I think that the hash method can only start part of the "lifecycle" if it is done automatically on the immutable state created on copy. However I don't think this will work because hashing is expensive and, at least on the consensus node, should not be performed on the same thread as the copy itself.

### Restricting Copying

To ensure that only `StateLifecycleManager` could manage mutability of the state object, a new implementation of the `State` interface must be introduced:

```java
public final class NonCopyableState implements State {
    private State delegate; 
    // All methods delegate to the delegate object except for the copy method.
    // For the copy method, we retain the default implementation, which throws an `UnsupportedOperationException`.
}
```

Client code requiring a mutable copy of the state must use the `copy` method provided by the `StateLifecycleManager` interface.

### State Initialization

The current state initialization logic is fragmented across multiple classes and methods, including:

- `ServicesMain#main`
- `Hedera#newMerkleStateRoot`
- `StartupStateUtils.getInitialState`
- `StartupStateUtils.loadStateFile`
- `StartupStateUtils.loadLatestState`
- `SignedStateFileReader.readStateFile`

Additionally, this logic is tightly coupled with other functionalities, such as:

- State signatures
- Platform service registration
- Invalid state file cleanup
- Hash validation
- State reservation

> Invalid state file cleanup and hash validation should be moved to the `StateLifecycleManager` implementation. The remaining responsibilities belong to the platform code.

Reviewer 2: What kind of "invalid" checks? Will these checks be common to both CN and BN?

Author: I added a clarification for the "invalid" checks. Yes, they will be common for CN and BN.

   - Providing the state for signing  
   - Initializing the state by extracting it from the signed state  

## Proposed Changes

All lifecycle-related methods from `com.swirlds.state.State` should be moved to a separate class responsible for managing the state lifecycle. This will provide a clearer separation of concerns.  

- `com.swirlds.state.State` should focus solely on accessing and modifying the state.  
- The state lifecycle will be managed externally.

The `SwirldStateManager` class requires refactoring to separate its two distinct responsibilities. `swirlds-state-impl` should not include any platform-specific details but should provide the necessary mechanisms for the platform to interact with the state and its lifecycle.

### `StateLifecycleManager` Interface

To achieve this, `swirlds-state-api` will define a `StateLifecycleManager` interface, with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`. The interface will handle:

- Creating/loading an initial state. If no snapshot is available, a new state will be created.  
- Managing references to the latest immutable state and the latest mutable state.

> - Resetting the mutable state to the latest immutable state.

Reviewer 3: I don't understand this resetting functionality, can you add a bit more detail please? During a reset, what happens to the mutable state and to the immutable state?

Author: > During a reset, what happens to the mutable state and to the immutable state?

The previous mutable state get discarded from the `StateLifecycleManager` (it's a client's code responsibility to release and destroy it). Current immutable state remains unchanged.

Reviewer: Exactly.  Block Node will need the ability to reset to the last immutable state and begin applying changes again from that point.
Examples for when this might be needed include
1. A block fails verification, so we have to request it be resent or request it from another source.
1. A stream is interrupted mid-block (e.g. by a network fault) and we need to request resend or from another source.

- Creating/loading an initial state. If no snapshot is available, a new state will be created.  
- Managing references to the latest immutable state and the latest mutable state.
- Resetting the mutable state to the latest immutable state.
- Restricting mutability to a single state object. Specifically, only one state can be mutable at a time.  
- Loading snapshots from disk. States loaded from snapshots are immutable.  

The interface will look something like this:

```java
public interface StateLifecycleManager {
    /**
     * Get the mutable state.
     */
    State getMutableState();

   /**
    * Reset the mutable state to the latest immutable state.
    */
   void resetMutableState();

   /**
    * Creates a snapshot for the latest immutable state. 
    * 
    * @param targetPath The path to save the snapshot.
    */
    State createSnapshot(final @NonNull Path targetPath);

    /**
     * Creates a mutable copy of the state. The previous mutable state becomes immutable,
     * replacing the latest immutable state. 
     * 
     * @return a mutable copy of the state
     */

>     State copyMutableState();

Reviewer 2: I think this should to return the immutable state that was used as the source of the copy operation. This is currently what the copy method in the `SwirldsStateManager` class returns because that is what is used for the signed state. It is the thing that moves forward in the lifecycle. If it is not returned by this method, the caller code will look something like this:

```
State soonToBeImmutableState = sslm.getMutableState();
sslm.copyMutableState();
SignedState ss = new SignedState(soonToBeImmutableState); // now is immutable
```

If it returns the immutable state, then it looks like this:

```
State completedState = sslm.copyMutableState();
SignedState ss = new SignedState(completedState);
```

Author: Yes, that's a good point. I'll make this change.

  as we normally do not allow creation of immutable state copies.
- The `createSnapshot` method requires the latest immutable state to be hashed. It will block on the `Future` returned by the `State#getHash()` method until the hash is ready.

---

### Implementation Details

The implementation will include the methods defined in the interface and two additional static utility methods:

```java
public class StateLifecycleManagerImpl implements StateLifecycleManager {
    /**
     * Load an immutable copy of a state from the snapshot at the specified path.
     *
     * @param sourcePath the path to the snapshot
     * @return the immutable state loaded from the snapshot
     */
    public static State loadSnapshot(Path sourcePath) {
        // implementation
    }

    /**
     * Returns a list of paths to snapshots discovered in the specified directory.
     *
     * @param sourcePath the path to the directory
     * @return a list of paths to snapshots
     */
    public static List<Path> discover(Path sourcePath) {
        // implementation
    }
}
```

#### Notes

> - The `StateLifecycleManager` implementation will maintain a **1-to-1 relationship** with the state. 

Reviewer 2: I don't understand the change to being a 1-1 relationship with the state. How will it ensure a single mutable state at a given time if there is one instance per state? The app would have to keep track of the mutable state on it's own, which is one of the reasons we wanted this lifecycle manager.

Author: I'll clarify the wording but what I meant is that this relationship is between the lifecycle manager and a sequence of states. It is **not** a relationship between one instance of the state and one instance of the lifecycle manager. 

Imagine a hypothetical scenario in which we need to support more than one network. We will to have one lifecycle manager per network, and the lifecycle manager will have two state references at a time (current mutable state and the latest immutable state). However, these references will be updated over time and will point to other instances.

The assumption is that the application will require only one mutable state at a time, and this state should be initialized during the application's startup.  

> - If client code needs a state loaded from an arbitrary snapshot, it can directly use the `StateLifecycleManagerImpl#loadSnapshot` method, which is `static`. 

Reviewer 2: The client code should have any access to the implementation class, only the interface, so I don't see how this will work

Author: The client code will need to use the implementation classes anyway. 

However, instead of having static method we can have a constructor that takes a path to the snapshot. And this is how the client code will be able to work with the loaded state.

# Lifecycle Management for the State

## Summary

> This proposal describes a possible implementation of state lifecycle management that could work equally well for both the Consensus Node and the Block Node.

Reviewer 4: ```suggestion
This proposal describes a possible implementation of state lifecycle management that could work equally well for the Consensus Node, Block Node and any clients that parse network streams and manage state.
```

|--------------------|------------------------------------------|
| Designers          | [@imalygin](https://github.com/imalygin) |
| Functional Impacts | Consensus Node, Block Node               |

## Purpose and Context

There is an initiative to create self-contained State modules (`swirlds-state-api` and `swirlds-state-impl`) that have minimal dependencies. Specifically, these modules should not depend on `swirlds-platform-core`. As part of this initiative, a state lifecycle management mechanism will be implemented for use by both the Consensus Node and the Block Node.

### Requirements

- Introduce an interface representing state lifecycle management and its implementation.
- Update the current State API to align with the State Lifecycle API.
- Implement these APIs and migrate existing code to use them.

## Design & Architecture

### Summary

The following design describes a solution for managing the lifecycle of the state object. It covers the following aspects:

- State initialization (loading a snapshot or creating a new state)
- Maintaining references to the mutable state and the latest immutable state
- Restricting mutability to a single state object

> - Loading snapshots from disk

Reviewer 4: This is a repetition of the first bullet point. Should this rather be persistence a snapshot to disk?
I think serializing a snapshot should be a feature

> - Loading snapshots from disk

Reviewer 4: Is there feature support for applying changes or is that the State API?

## Current State of the Code

> ### Reservation Count Mechanism

Reviewer 5: I really want this to go away. It just feels over complicated. The basic problem is we have a chain of versions starting from the mutable version going back to the oldest known version on disk. We have to explicitly delete a version when we are finished with it so it can be merged in virtual node cache or written to disk and freed in memory. If we make it explict manual in State API to have to call release/delete(version) then the state machine that controls states will have to keep track of if it needs a state or not and call release() when done. This will normally be at the end of processing and writing of blockStream block. But sometimes one will be held longer for state saving. Or coping to teach from. 

Maybe we would have to do this later when we can delete all non-virtual-map merkle nodes as will be very hard to change before that.

is presumed to still need the object. The reservation count for an object with an implicit reservation is 0.
If `Releasable#release()` is called on such an object, it will be destroyed.

#### Explicit Reservations

When an object with an implicit reservation has `Reservable#reserve()` called on it for the first time, the reservation
becomes explicit, and its reservation count increases to 1. From this point forward, the reservation remains
explicit until the object is destroyed. An object with an explicit reservation cannot return to having an implicit reservation.

If, at any time after obtaining an explicit reservation, the object has `Releasable#release()` called enough times
to reduce its reservation count to 0, the object will be destroyed, and its reservation count will be set to -1.

---

We rely on this mechanism to manage the number of states kept in memory.

**Important**: It is the responsibility of the client code to ensure that the state is released once it is no longer needed.

Once the reservation count reaches 0, the state becomes eligible for garbage collection.

---

### Java Classes

`com.swirlds.state.State` provides the following lifecycle-related methods:

> - `State copy()`: Creates a mutable copy of the state. The source state becomes immutable.

Reviewer 5: Don't love the name `copy()` as it implies it is completely separate copy of tree. What this really is doing is creating a new version so maybe calling it `newVersion()` is more obvious.

Author: >Don't love the name copy() as it implies it is completely separate copy of tree. 

I was following the pattern that we have for `MerkleNode`, and, correct me if I'm wrong, but it doesn't imply a completely separate copy of the tree. Also, it follows the pattern we use in the other parts of the code - once we create a copy, the source becomes immutable.

> - `void computeHash()`: Computes the hash of the immutable state.

Reviewer 6: Compute the state hash. Can only be called on immutable states

Reviewer 4: For information reason why can it only be called on immutable state?
Is there consideration for being able to compute the hash in the mutable state?
Say I start with State A, do a `copy()`, apply state changes to the mutable state and want to capture the hash of the state at that point.
Is this feasible and supported?

> - `void createSnapshot(Path targetPath)`: Creates a snapshot of the state at the specified path.

Reviewer 4: Same note above. Why only on immutable. What about capture the point in time after a select number of changes and persisting that?

- `State loadSnapshot(Path sourcePath)`: Loads a snapshot from the specified path.

> 

Reviewer 5: should there be a `createNewState()` method to create a brand new state?

Author: That would be up to `StateLifecycleManager`. It will be responsible for either loading an existing snapshot or for creating an instance of a new state.

Reviewer 6: I am slightly confused about loadSnapshot(). As far as I can see, a State instance is to be created first, then loadSnapshot() is called for this instance, is it correct? Shouldn't it be a factory method somewhere else, e.g. StateLifecycleManager?

Reviewer 4: My reading of this is loadSnapshot() would load state into the immutable version.
So maybe to start you would either loadSnapshot() or create a blank state to start with.
@imalygin is that what you were going for?

If so then similar to @jasperpotts comment one approach might be to utilize a create method.
Maybe it's `create()` & `create(Path sourcePath)`, with the later loading a snapshot from disk.
Or just a `create(Path sourcePath)` where if `sourcePath` is `null` you create an empty state to start with

Reviewer 4: `createImmutable(...)` is also another option as it then allows you to rename `copy()` to `createMutable()` which highlights the need for an immutable state first and then you copy from immutable to Mutable state

Author: @Nana-EC not exactly. `StateLifecycleManager` should be responsible for creation of snapshot and loading state from them.

These methods correspond to the state lifecycle as follows:

[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)

---

> ### `SwirldStateManager`

Reviewer 3: I think there is a PR to delete this class

   - Creating copies of the state and updating references.
     This functionality belongs in `swirlds-state-api`.
2. **Platform-Related Event Handling** (belongs to `swirlds-platform-core`):
   - Handling consensus rounds
   - Sealing consensus rounds
   - Keeping the `freeze time` parameter up to date
   - Providing the state for signing
   - Initializing the state by extracting it from the signed state

## Proposed Changes

All lifecycle-related methods from `com.swirlds.state.State` should be moved to a separate class responsible for managing the state lifecycle. This will provide a clearer separation of concerns.

- `com.swirlds.state.State` should focus solely on accessing and modifying the state.
- The state lifecycle will be managed externally.
- Its hash will be computed asynchronously and managed externally.

```java
public interface State {
   @NonNull
   ReadableStates getReadableStates(@NonNull String serviceName);

   @NonNull
   WritableStates getWritableStates(@NonNull String serviceName);

   default void registerCommitListener(@NonNull final StateChangeListener listener)l;

   @Nullable

>    Future<Hash> getHashFuture();

Reviewer 6: computeHash(), maybe?

The `SwirldStateManager` class requires refactoring to separate its two distinct responsibilities. `swirlds-state-impl` should not include any platform-specific details but should provide the necessary mechanisms for the platform to interact with the state and its lifecycle.

### `StateLifecycleManager` Interface

To achieve this, `swirlds-state-api` will define a `StateLifecycleManager` interface, with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`. The interface will handle:

- Creating/loading an initial state. If no snapshot is available, a new state will be created.
- Managing references to the latest immutable state and the latest mutable state.
- Resetting the mutable state to the latest immutable state.
- Restricting mutability to a single state object. Specifically, only one state can be mutable at a time.
- Loading snapshots from disk. States loaded from snapshots are immutable.

The interface will look something like this:

```java
public interface StateLifecycleManager {
    /**
     * Get the mutable state.
     */
    State getMutableState();

   /**
    * Reset the mutable state to the latest immutable state.
    */

>    void resetMutableState();

Reviewer: This was discussed in an [earlier conversation](https://github.com/hiero-ledger/hiero-consensus-node/pull/16592#discussion_r1890840031) on this PR.

| Designers          | [@imalygin](https://github.com/imalygin) |
| Functional Impacts | Consensus Node, Block Node                 |

## Purpose and Context

There is an initiative to create self-contained State modules (`swirlds-state-api` and `swirlds-state-impl`) that would have minimal dependencies, and specifically, they should not depend on `swirlds-platform-core`. Part of this initiative is the implementation of a state lifecycle management mechanism that could be used by both the Consensus Node and the Block Node.

### Requirements

- Add an interface representing state lifecycle management and its implementation.
- Update the current State API to be coherent with this State Lifecycle API.
- Implement this set of APIs and migrate the existing code to use it.

## Design & Architecture

### Java classes

- `com.swirlds.state.State` has the following lifecycle-related methods:
  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
  - `void computeHash()`: computes the hash of the immutable state.
  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.

>   - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.

Reviewer 2: I think it can be static, at least for now. When we get to supporting multiple states from different state families in memory at the same time, we may decide to change it.

>   - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.

Author: As of now, this method is not static. Kelly is right, technically it can be static. However, it would be suboptimal because we'd have to move `MerkleTreeSnapshotReader` class to `swirlds-state-api`. As the name suggest, it knows about the implementation details of the `State` and therefore it doesn't belong to `swirlds-state-api`. 

I'd like to move this method to the `StateLifecycleManager` because we don't want this logic to require an instance of `State` object.

>   - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.

Reviewer 2: Should this method still be here then? I agree, it should be in another class. `StateLifecycleManager` is good.

>   - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.

Author: It should not be there and it should be static and be a part of `StateLifecycleManagerImpl`

These methods correspond to the state lifecycle as follows:

[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)

Note that the `loadSnapshot` method doesn't really belong to `com.swirlds.state.State`, as it's a method of the class, and in some cases, we may need to load the snapshot before we have an instance of the state. 
This method should be moved to a separate class that will be responsible for managing the state lifecycle.

> - `SwirldStateManager` needs refactoring. The current implementation has two sets of responsibilities:

Author: >The Swirld prefix (e.g. SwirldState, SwirldStateManager) is confusing. My understanding is it actually means "platform". It would be great to rename the classes accordingly in the future

Yes, I agree. Whenever I move the functionality to the `swirlds-state-api` module, I drop this prefix. Generally, I agree - we should get rid of this prefix in the `swirlds-platform-core` as well

>   - Keeping track of the references to the latest immutable state and the latest mutable state. This includes creating a copy of the state and updating the references.

Reviewer 6: This responsibility and related methods should be a part of `swirlds-state-api` module. The other one will be kept in `swirlds-platform-core`

Author: Sure. I'll mention that explicitly.

  - Handling Platform-related events:
    - Handling consensus rounds.
    - Sealing consensus rounds.
    - Keeping the `freeze time` parameter up-to-date.
    - Providing state for signing.
    - Initializing the state by extracting it from the signed state.

These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should provide all necessary levers for the platform to interact with the state and its lifecycle.

> Therefore, `swirlds-state-impl` needs a `StateManager` class that will have the following responsibilities:

Reviewer 6: Does it make sense to have a state manager interface in the `swirlds-state-api` module, and one more class that implements the interface in `swirlds-state-impl`?

Author: The idea was to have an implementation only, but you're right - we'd better have `StateManager` as an interface and `StateManagerImpl` as an implementation

- Keeping track of the references to the latest immutable state and the latest mutable state.
- Tracking states that existed before the latest immutable state. These should be available for lookup by their round number.
- Evicting old states that are no longer needed.
- Loading snapshots from the disk.

Additionally, there should be a configuration parameter that limits the number of states kept in memory. When the limit is reached, the oldest state should be evicted.

> The interface should look as follows:

Reviewer 2: Something we've learned in these proposals is that we should be flexible with the exact APIs. I suggest changing this sentence to this to make it more flexible so that if something small changes, you don't need to come back and change this section:
```suggestion
The interface will look something like this:
```

Author: :) fair enough. I'll rephrase it

```java
public interface StateManager {
    State getLatestImmutableState();
    

>     State getLatestMutableState();

Reviewer 6: It can be just `getMutableState()`, since there is only one mutable state and it's always the latest.

Author: Agreed

    
    State getState(long round);
    

>     void evictState(long round);

Reviewer 6: This one is questionable... In the current implementation, states are released using ref counting mechanism. Will it be different in `swirlds-state-impl`?

Reviewer 2: Yes, we want to get rid of the reference counting all together. The state module will retain a state in memory until it is told to release it, or possibly until it reached the configured maximum age, although i'm not sure about this second part.

Author: I'll explicitly mention that `evictState` is there as a replacement for the reference counting mechanism

    
    State loadSnapshot(Path sourcePath);
}
```

> Note that the loaded snapshot should be treated as one of the immutable states and should be available via `getState`.

Author: >Should a call to evictState() on the mutable state should fail?

Yes, absolutely. I can't imagine a scenario where it would be appropriate. 

>How will reconnect work with this lifecycle manager?

I think it should work similarly to how it works now. @artemananiev  correct me if I'm wrong, but here is how I understand it: 
1. We load the latest snaphot
2. We create a copy of it to get a mutable instance
3. We apply a set of changes to the state as a part of the reconnect process. 

If I'm not mistaken, we don't need to set the state explicitly from the outside. `StateManager` can take care of it.

> Note that the loaded snapshot should be treated as one of the immutable states and should be available via `getState`.

Reviewer 2: Yes, you are correct for how it works. I was not thinking about it correctly.

This method should be moved to a separate class that will be responsible for managing the state lifecycle.

`SwirldStateManager` needs refactoring. The current implementation has two sets of responsibilities:
  - Keeping track of the references to the latest immutable state and the latest mutable state. This includes creating a copy of the state and updating the references. 
  This functionality belongs to `swirlds-state-api`.   
  
  - Handling Platform-related events (belongs to `swirlds-platform-core`):
    - Handling consensus rounds.
    - Sealing consensus rounds.
    - Keeping the `freeze time` parameter up-to-date.
    - Providing state for signing.
    - Initializing the state by extracting it from the signed state.

These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should provide all necessary levers for the platform to interact with the state and its lifecycle.

Therefore, `swirlds-state-api` needs a `StateManager` interface with an implementation `StateManagerImpl` in `swirlds-state-impl`  that will have the following responsibilities:

> - Genesis state initialization.

Reviewer 2: ```suggestion
- Create an empty state
```

- Keeping track of the references to the latest immutable state and the latest mutable state.
- Tracking states that existed before the latest immutable state. These should be available for lookup by their round number.

> - Restricts mutability of the state to a single object. That is, only one state object should be mutable only once.

Reviewer 2: ```suggestion
- Restricts mutability to a single state. That is, only one state object should be mutable at any given time.
```

> - Evicting old states that are no longer needed. `evictState` method *must* fail if the state is still mutable.

Reviewer 2: ```suggestion
- Evicting old states from memory that are no longer needed. `evictState` method *must* fail if the state is still mutable.
```

- Loading snapshots from the disk.

Additionally, there should be a configuration parameter that limits the number of states kept in memory. When the limit is reached, the oldest state should be evicted.
Currently, we rely on the implicit garbage collection. We have a reference count mechanism to achieve this.
We want to get rid of this mechanism and rely on the explicit eviction mechanism.

> The following diagram illustrates the proposed illustrate management:

Reviewer 2: ```suggestion
The following diagram illustrates an example of the states in memory throughout their lifecycles:
```

This functionality belongs to `swirlds-state-api`.

- Handling Platform-related events (belongs to `swirlds-platform-core`):
  - Handling consensus rounds.
  - Sealing consensus rounds.
  - Keeping the `freeze time` parameter up-to-date.
  - Providing state for signing.
  - Initializing the state by extracting it from the signed state.

These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should provide all necessary levers for the platform to interact with the state and its lifecycle.

Therefore, `swirlds-state-api` needs a `StateLifecycleManager` interface with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`  that will have the following responsibilities:
- Take initial state as a starting point. It can be a genesis state or a state loaded from a snapshot.
- Keep track of the references to the latest immutable state and the latest mutable state.
- Restrict mutability to a single state. That is, only one state object should be mutable at any given time.

The interface will look something like this:

```java
public interface StateLifecycleManager {

>     void setInitialState(State initialState);

Reviewer 2: In the normal startup of a CN that has a state on disk, will this be how the state lifecycle manager gets initialized?

1. Load state from disk (I think this should be in the StateLifecycleManager, explained in a [comment](https://github.com/hashgraph/hedera-services/pull/16592/files/160388010a7ac16edfd958e431815dd40bb26080..5b058e400180e8d1a7fcc8c032de6f2072b56b7d#r1853936664) below
2. Use that state to invoke `setInitialState()` which then performs a fast copy, setting both the latest immutable state and the mutable state

Author: Yes, this makes sense. In this case there is no need to expose `setInitialState`, as it should be a responsibility of the `StateLifecycleManager` to deal with it

Author: I'll update the doc accordingly

    State getLatestImmutableState();

    State getMutableState();
}
```

**Imporatant**: It's up to the client code to make sure that the state is released once it's no longer needed.

> Without this, the state will not be evicted from memory and eventually the application will run out of memory.

Reviewer 2: Two comments:
1. Typo in "Important"
2. Nothing about reservations has been explained in this document so far so this note is not meaningful. You don't need to go into the details of how it works, but a short paragraph explaining that reference counts are used to determine when a state is eligible for GC would be helpful. Then you could also state that the state module will ensure the mutable and latest immutable states are held onto, but if the user of the library needs to keep a state around longer than that, it is their responsibility to ensure proper use of additional reservations. The CN will do with with the existing class structures of SignedState and ReservableSignedState that sit on top of these state reservations.

Author: Sure, I will add information about reservations.

The following diagram illustrates an example of the states in memory throughout their lifecycles:

[![Multistate management](multi-states.svg)](multi-states.svg)

Also, `swirlds-state-api` needs a `SnapshotManager` interface with an implementation `SnapshotManagerImpl` in `swirlds-state-impl` that will have the following responsibilities:

> - Load snapshots from the disk. A state loaded from a snapshot must be immutable and available for look up by the round number.

Reviewer 2: ```suggestion
- Load snapshots from the disk. A state loaded from a snapshot is immutable and available for look up by the round number.
```
The method to get a state by round number was removed which I think is correct. This verbage needs to be updated to say that the snapshot loaded from disk will be fast copied. The immutable version loaded from disk will be available via the call to `getLatestImmutableState()` and the mutable copy will be available via `getMutableState()`

> - Evict the loaded state by the round number when the state is no longer needed.

Reviewer 2: This has been removed, so this line should also be removed

> - Evict the loaded state by the round number when the state is no longer needed.

Reviewer 2: I don't think we need this anymore

Author: Yes, but we need self-management for the number of states in memory. I'll add more details.

This functionality is separated from the `StateLifecycleManager` for the following reasons:
- a state loaded from a snapshot is _always_ immutable, and therefore there is no notion of mutability here.
- the mechanism of maintaining the loaded states in memory is different. This functionality doesn't rely on the reference counting.
Instead, it has explicit methods to load a state from a snapshot, and evict the state by the round number when it's no longer needed.

The interface will look something like this:

```java
public interface SnapshotManager {

    State getLoadedState(long round);

    State loadSnapshot(Path sourcePath);

    void evictLoadedState(long round);

> }

Reviewer 2: I don't agree with the splitting of methods this way. The idea of the snapshot manager was to deal with instances of `Snapshot` which will be a new class that loads minimal information about a state from disk like the hashes that can be used to produce a merkle proof, but not load the entire state into memory. Loading an entire state into memory and providing it as a `State` object is a function of a `StateManager` or `States` object.

The `loadSnapshot` method should probably be `loadStateFromDisk` to differentiate loading the minimal data from loading the full state into memory.

Also, I don't see a reason for the `evict` or `getLoadedState` method anymore. When a state's references go to zero, it will should be GCed. And if the user of the module need to hang onto a state for collecting signatures or anything like that, it's going to have to keep track of the reference itself to prevent it from being GCed, therefore this method is not needed.

Author: Right now we don't maintain a list of the loaded states. There is no such functionality - we don't have a class that has `State getLoadedState(long round)` method. If we intend to have such a method, there should be some way to get rid of the states that are no longer needed, manual or automatic.

```

### Metrics

> We need to monitor the number of states that are kept in memory by `SnapshotManager` to ensure there is no memory leak.

Reviewer 2: I think this can be removed

Author: `StateLifecycleManager` will maintain a list of states previously loaded with `loadSnapshot` method. We need to monitor the size of this list to make sure that we don't have unused state accumulating in memory or they may be all in use but it's just too many of them which may cause OOM. 

Alternatively, we can get rid of `State getLoadedState(long round)` method and use `loadSnapshot` whenever we need to access a state on the disk.

### Performance impact

> There is only one important performance consideration—the number of states that are kept in memory. However, this can be mitigated by the configuration parameter that limits the number of states kept in memory. In the absence of this parameter, it should be carefully tested and monitored to prevent memory leaks.

Reviewer 2: I think this part is now obsolete
